src/lua/compat.lua

```diff
@@ -0,0 +1,339 @@
+                                       and action function
+]]
+
+-- Returns table with all non-obsolete options from `options` and their values.
+local function serialize_compat()
+    -- The results of serialization should be ordered correctly. The only
+    -- feasible way for it is using indexed table with values {option_name = val}.
+    local result = { }
+
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if option.selected and option.current == NEW and not option.obsolete then
+            table.insert(result, {[name] = 'new'})
+        end
+    end
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if option.selected and option.current == OLD and not option.obsolete then
+            table.insert(result, {[name] = 'old'})
+        end
+    end
+    for _, name in pairs(options_order) do
+        local option = options[name]
+        if not option.selected and not option.obsolete then
+            table.insert(result, {[name] = ('default (%s)'):format(option.current and 'new' or 'old')})
+        end
+    end
+
+    -- Result is an indexed table but this way user can interact with it as if
+    -- it was key-value.
+    result = setmetatable(result, {
+        __index = function(self, key)
+            for i = 1, #self do
+                if self[i][key] ~= nil then
+                    return self[i][key]
+                end
+            end
+        end
+    })
+
+    return result
+end
+
+-- Checks options correctness.
+local function verify_option(name, option)
+    if type(name) ~= 'string' then
+        error(('Option name must be a string (%s provided)'):format(type(name)))
+    end
+    for p, t in pairs(options_format) do
+        if not t:find(type(option[p])) then
+            error(('Invalid option table for %s, bad %s (%s is expected)'):format(name, p, t))
+        end
+    end
+    if option.default ~= 'new' and option.default ~= 'old' then
+        error(("Invalid option table for %s, bad default ('new'/'old' is expected)"):format(name))
+    end
+    if option.obsolete and option.default == 'old' then
+        error(('Obsolete option %s default is wrong.'):format(name))
+    end
+end
+
+-- Checks if operation is valid, sets value to an option and runs postaction.
+local function set_option(name, val)
+    local option = options[name]
+    if not option then
+        error(('Invalid option %s'):format(name))
+    end
+    local selected
+    if val == 'new' then
+        val = NEW
+        selected = true
+    elseif val == 'old' then
+        val = OLD
+        selected = true
+    elseif val == 'default' then
+        val = option.default == 'new'
+        selected = false
+    else
+        error(('Invalid argument %s'):format(val))
+    end
+    if option.obsolete and val == OLD then
+        error('Chosen option is no longer available')
+    end
```

Comment:
Nit: I think an option name would be helpful for a user.

Reply:
Done.